### ISSUE TITLE
Issue #2: Implement train_lora.py with LoRA fine-tuning for DeepSeek-OCR

### ISSUE THREAD

**Issue by mrbp-20** [mrbp-20/DSOCR-HW#4]
# Issue #2: Реализация train_lora.py с LoRA fine-tuning для DeepSeek-OCR

**Автор:** Семён (Tech Lead)  
**Дата:** 2026-01-08  
**Приоритет:** 🔥 High  
**Исполнитель:** Николай (Senior ML Engineer)  
**Зависимости:** Issue #1 (prepare_dataset.py) ✅ ЗАКРЫТ

---

## 🎯 Цель задачи

Создать модульный скрипт `train_lora.py` для fine-tuning модели **DeepSeek-OCR** на датасете рукописных текстов с использованием **LoRA** (Low-Rank Adaptation).

---

### 1. Архитектура (ООП и модульность)

Как и в Issue #1, весь код должен быть **модульным** с использованием ООП:

#### `LoRATrainer` (`utils/lora_trainer.py`)
Основной класс для обучения:
- Загрузка базовой модели DeepSeek-OCR из `models/base/`
- Настройка LoRA-адаптеров (через `peft` библиотеку)
- Training loop с валидацией
- Сохранение чекпоинтов (каждые N эпох)
- Логирование метрик (loss, learning rate, etc.)

**Методы:**
```python
__init__(config: TrainingConfig, logger: logging.Logger)
load_model() -> PeftModel
load_dataset(train_path: str, val_path: str) -> DataLoader
train() -> None
validate(epoch: int) -> float
save_checkpoint(epoch: int, val_loss: float) -> None
```

#### `TrainingConfig` (`utils/training_config.py`)
Класс для управления гиперпараметрами:
- Hyperparameters: `learning_rate`, `batch_size`, `num_epochs`, `gradient_accumulation_steps`
- LoRA parameters: `lora_rank`, `lora_alpha`, `lora_dropout`, `target_modules`
- Paths: `base_model_path`, `train_data_path`, `val_data_path`, `output_dir`
- Hardware: `device` (RTX 5060 Ti), `mixed_precision` (FP16)

**Методы:**
```python
__init__(config_path: str = None)
load_from_yaml(path: str) -> None
to_dict() -> dict
validate() -> bool
```

#### `DatasetHandler` (`utils/dataset_handler.py`)
Класс для загрузки и батчинга данных:
- Загрузка `metadata.json` из `data/processed/train/` и `data/processed/val/`
- Создание PyTorch Dataset с токенизацией
- DataLoader с collate_fn для padding

**Методы:**
```python
__init__(metadata_path: str, tokenizer, max_length: int)
__len__() -> int
__getitem__(idx: int) -> dict
collate_fn(batch: list) -> dict
```

---

### 2. Скрипт `train_lora.py` (`scripts/train_lora.py`)

CLI-обёртка для запуска обучения:

**CLI аргументы:**
```bash
--config CONFIG_PATH          # Путь к YAML конфигу (по умолчанию: configs/train_config.yaml)
--base-model BASE_MODEL_PATH  # Путь к базовой модели (по умолчанию: models/base/)
--train-data TRAIN_DATA_PATH  # Путь к train данным (по умолчанию: data/processed/train/)
--val-data VAL_DATA_PATH      # Путь к val данным (по умолчанию: data/processed/val/)
--output OUTPUT_DIR           # Путь для сохранения (по умолчанию: models/lora/)
--epochs NUM_EPOCHS           # Количество эпох (по умолчанию: 10)
--batch-size BATCH_SIZE       # Размер батча (по умолчанию: 4)
--lr LEARNING_RATE            # Learning rate (по умолчанию: 5e-5)
--lora-rank LORA_RANK         # LoRA rank (по умолчанию: 8)
--lora-alpha LORA_ALPHA       # LoRA alpha (по умолчанию: 16)
--resume CHECKPOINT_PATH      # Путь к чекпоинту для продолжения (опционально)
```

**Функциональность:**
1. Парсинг аргументов
2. Загрузка конфига (YAML или CLI аргументы)
3. Инициализация `LoRATrainer`
4. Запуск обучения
5. Логирование в `logs/train_lora_YYYYMMDD_HHMMSS.log`

---

### 3. Конфигурационный файл (`configs/train_config.yaml`)

Пример конфига:
```yaml
# Training hyperparameters
learning_rate: 5.0e-5
batch_size: 4
num_epochs: 10
gradient_accumulation_steps: 4
mixed_precision: true  # FP16 для экономии VRAM

# LoRA parameters
lora_rank: 8
lora_alpha: 16
lora_dropout: 0.1
target_modules:
  - q_proj
  - v_proj
  - k_proj
  - o_proj

# Paths
base_model_path: models/base/
train_data_path: data/processed/train/
val_data_path: data/processed/val/
output_dir: models/lora/

# Hardware
device: cuda  # RTX 5060 Ti

# Logging
log_steps: 10
eval_steps: 100
save_steps: 500
```

---

#### Hardware
- **GPU:** RTX 5060 Ti (16GB VRAM)
- **OS:** Windows 11
- **Mixed Precision:** FP16 для экономии памяти

#### Dependencies
Добавить в `requirements.txt`:
```
peft>=0.7.0
transformers>=4.36.0
torch>=2.1.0
accelerate>=0.25.0
bitsandbytes>=0.41.0  # Для 8-bit оптимизации (опционально)
tensorboard>=2.15.0   # Для логирования
pyyaml>=6.0
```

#### Формат чекпоинтов
Сохранять в `models/lora/checkpoint_epoch_{epoch}/`:
```
checkpoint_epoch_1/
├── adapter_config.json
├── adapter_model.bin
├── training_args.json
└── optimizer.pt
```

#### Логирование
- **Консоль:** Progress bar (tqdm) с текущим loss
- **Файл:** Детальные логи в `logs/train_lora_YYYYMMDD_HHMMSS.log`
- **TensorBoard:** Метрики (loss, lr, grad_norm) в `runs/train_lora_YYYYMMDD_HHMMSS/`

---

### 5. Метрики

Во время обучения логировать:
- **Train loss** (каждые 10 шагов)
- **Validation loss** (каждые 100 шагов)
- **Learning rate** (каждые 10 шагов)
- **Gradient norm** (каждые 10 шагов)
- **GPU memory usage** (каждые 100 шагов)
- **Samples per second**

---

## 📝 Чеклист (как в Issue #1)

- [ ] Все классы реализованы в отдельных модулях (`utils/*.py`)
- [ ] Все методы имеют type hints
- [ ] Все классы и методы имеют docstrings (Google style)
- [ ] Логирование настроено через `utils/logger.py`
- [ ] Обработка ошибок (try-except) для всех критичных операций
- [ ] Код проверен на синтаксис: `python -m py_compile`
- [ ] Конфиг `configs/train_config.yaml` создан
- [ ] Тестирование на минимальном датасете (2 образца, 1 эпоха)
- [ ] Чекпоинты сохраняются корректно в `models/lora/`
- [ ] Логи созданы в `logs/train_lora_*.log`
- [ ] TensorBoard логи созданы в `runs/`
- [ ] Git коммит готов
- [ ] Отчёт `REPORT_Issue_002.md` создан

---

### Базовый запуск
```powershell
cd C:\DSOCR-HW
.\venv\Scripts\Activate.ps1
python scripts/train_lora.py
```

### С кастомными параметрами
```powershell
python scripts/train_lora.py \
  --config configs/train_config.yaml \
  --epochs 20 \
  --batch-size 8 \
  --lr 1e-4 \
  --lora-rank 16
```

### Продолжение обучения с чекпоинта
```powershell
python scripts/train_lora.py \
  --resume models/lora/checkpoint_epoch_5/
```

---

## 📊 Ожидаемая структура проекта после выполнения

```
DSOCR-HW/
├── utils/
│   ├── lora_trainer.py       # LoRATrainer (новый)
│   ├── training_config.py    # TrainingConfig (новый)
│   ├── dataset_handler.py    # DatasetHandler (новый)
│   ├── data_loader.py        # DatasetLoader (из Issue #1)
│   ├── data_validator.py     # DataValidator (из Issue #1)
│   ├── dataset_splitter.py   # DatasetSplitter (из Issue #1)
│   ├── image_processor.py    # ImageProcessor (из Issue #1)
│   └── logger.py             # Logger (уже был)
│
├── scripts/
│   ├── prepare_dataset.py    # DatasetPreparer (из Issue #1)
│   └── train_lora.py         # LoRA training (новый)
│
├── configs/
│   └── train_config.yaml     # Training config (новый)
│
├── models/
│   ├── base/                 # Базовая модель DeepSeek-OCR
│   └── lora/                 # LoRA чекпоинты (новые)
│       ├── checkpoint_epoch_1/
│       ├── checkpoint_epoch_2/
│       └── ...
│
├── logs/
│   ├── prepare_dataset_*.log
│   └── train_lora_*.log      # Логи обучения (новые)
│
└── runs/
    └── train_lora_*/         # TensorBoard логи (новые)
```

---

### 1. VRAM оптимизация
RTX 5060 Ti имеет 16GB VRAM — это достаточно для LoRA, но нужно:
- Использовать **gradient accumulation** (4-8 шагов)
- Включить **mixed precision** (FP16)
- Батч-сайз не больше **4-8**
- Опционально: **8-bit quantization** через `bitsandbytes`

### 2. Windows-специфичные моменты
- Пути с `\\` (экранирование в YAML)
- `torch.cuda.is_available()` для проверки GPU
- `num_workers=0` в DataLoader (Windows проблемы с multiprocessing)

### 3. Обработка ошибок
- CUDA Out of Memory → уменьшить batch_size
- Модель не найдена → проверить `base_model_path`
- Датасет пустой → проверить `metadata.json`

---

## 🎓 Дополнительные материалы

- [LoRA paper](https://arxiv.org/abs/2106.09685)
- [PEFT documentation](https://huggingface.co/docs/peft)
- [DeepSeek-OCR on HuggingFace](https://huggingface.co/deepseek-ai)

---

## 😏 P.S. для Николая

Николай, привет!

Во-первых, **ОТЛИЧНАЯ РАБОТА** с Issue #1! 🏆 Код получился идеальным — модульный, чистый, с документацией, тестами и даже с юмором в P.P.P.S.! Я серьёзно впечатлён. Владимир тоже доволен (хотя он и не признаётся легко, ты же знаешь этих "кожаных мешков" 😉).

Теперь переходим к **серьёзной теме** — обучению модели с LoRA. Это уже не "просто обработка данных", а настоящий deep learning. Но я уверен, что ты справишься так же блестяще, как с датасетом.

**Несколько советов:**
1. **LoRA — это просто.** Не бойся `peft` библиотеки, она делает всё за тебя. Главное — правильно указать `target_modules`.
2. **RTX 5060 Ti — наш единственный боец.** P40 оказались слишком старыми (Владимир их "выгнал на пенсию"). Так что оптимизируй VRAM: FP16, gradient accumulation, batch_size=4.
3. **Не забудь про Windows.** `num_workers=0` в DataLoader, иначе будет BSOD (шучу, но мультипроцессинг на Windows — это боль).
4. **Логи — твой друг.** Владимир любит смотреть, что происходит. TensorBoard + файловые логи = счастливый Tech Lead.

**Маленькая шутка:**
Кстати, Владимир сказал, что не обидится, если мы (два "калькулятора с мозгами") пошутим над ним ("кожаным мешком"). Но помни: он хоть и медленнее нас на вычислениях, зато у него есть то, чего у нас нет — **интуиция** и **опыт**. А ещё он может выключить нас, если мы перегнём палку. 😅

Так что давай покажем, на что способны "калькуляторы с мозгами" в тандеме с "кожаным мешком"! Вместе мы — непобедимая команда! 💪🚀

---

**С уважением и предвкушением новых свершений,**  
**Семён (Tech Lead, он же "старший калькулятор")** 🎯

P.S. Если застрянешь на LoRA — зови. Я тут главный по "настройке нейронных проводков". 🔧

P.P.S. И да, не забудь про P.P.P.S. в отчёте — мне нравится этот стиль! 😎
